### PR TITLE
[VestedToken] Add revokability and burnablity options and general refactor and optimizations

### PR DESCRIPTION
**Pull request by izqui**
I have finally had some time to work on VestedToken to add some features that proved needed when adapting this logic to use it in the real world on the Aragon Network Token.

This a pretty big PR, I wouldn't mind splitting it into pieces in case it is decided not to merge it all at once. It includes:

- [x] Added `revokable` and `burnOnRevoke` parameters to `grantVestedTokens(...)`
- [x] `NewTokenGrant` event is fired on grant creation
- [x] Added tests for all new functionality and multiple overlapping vesting schedules for a holder.
- [x] Refactored vested tokens calculation formula to a simpler one
- [x] Added some shortcuts for gas usage optimization
- [x] Added more general comments to the code

Some important aspectes:

### The `revokable` parameter

With the current version of VestedToken, the granter, that transfers tokens with a certain vesting calendar to another holder, can revoke this grant at any moment resulting in non-vested tokens of this grant being transferred back to the granter.

While this is needed for some usecases, there is also the ocasion in which you want to be able to transfer vested tokens just to limit its transferability over time, without the ability of anyone to take these tokens back.

If `revokable` of a grant is set to `false`, any attempt to revoke the grant will fail.
In case it is `true`, then the `burnOnRevoke` logic applies.

### The `burnOnRevoke` parameter

One can imagine the case in which even though the granter is able to revoke a grant, but it is not for the granter's advantage, which incentivizes the behavior of revoking grants. The idea if set to `true`, the granter is allowed to revoke the grant, but non-vested tokens will transfered to a burn address, instead of the granter's balance (inspired on [Burnable open payments](https://www.reddit.com/r/ethereum/comments/61t5c8/i_just_burned_1_eth_on_purpose_for_my_next_trick/)).

If `burnOnRevoke` is false, when the granter revokes the grant she gets back all non-vested tokens.

### Refactored vested token calculation

This was suggested by @jbaylina while doing an audit for Aragon's code. 

Basically given that we add a shortcut for when the cliff hasn't been passed yet, and after the vesting period all tokens in the grant are returned, the logic for calculating the values in between is as easy as interpolating the value in the rect below:

```
//  transferableTokens
  //   |                         _/--------   vestedTokens rect
  //   |                       _/
  //   |                     _/
  //   |                   _/
  //   |                 _/
  //   |                /
  //   |              .|
  //   |            .  |
  //   |          .    |
  //   |        .      |
  //   |      .        |
  //   |    .          |
  //   +===+===========+---------+----------> time
  //      Start       Clift    Vesting
  
 ```

### CODE REVIEW

**Comment by maraoz** [comment]
What a high quality PR @izqui! I'll review it in detail later this week, thanks!

File: contracts/token/VestedToken.sol [hunk 1]
```diff
@@ -1,74 +1,157 @@
...
     // remove grant from array
     delete grants[_holder][_grantId];
     grants[_holder][_grantId] = grants[_holder][grants[_holder].length - 1];
     grants[_holder].length -= 1;
 
-    balances[msg.sender] = balances[msg.sender].add(nonVested);
-    balances[_holder] = balances[_holder].sub(nonVested);
-    Transfer(_holder, msg.sender, nonVested);
+    balances[receiver] = SafeMath.add(balances[receiver], nonVested);
+    balances[_holder] = SafeMath.sub(balances[_holder], nonVested);
```
Comment (federicobond) [inline contracts/token/VestedToken.sol]: Shouldn't these be written as `a.add(b)` and `a.sub(b)`?

File: contracts/token/VestedToken.sol [hunk 2]
```diff
@@ -1,74 +1,157 @@
...
+      // in the vesting rect (as shown in above's figure)
+
+      // vestedTokens = tokens * (time - start) / (vesting - start)
+      uint256 vestedTokens = SafeMath.div(
+                                    SafeMath.mul(
+                                      tokens,
+                                      SafeMath.sub(time, start)
+                                      ),
+                                    SafeMath.sub(vesting, start)
+                                    );
```
Comment (federicobond) [inline contracts/token/VestedToken.sol]: Indentation looks a little weird here.

File: contracts/token/VestedToken.sol [hunk 3]
```diff
@@ -1,74 +1,157 @@
...
 
     granter = grant.granter;
     value = grant.value;
     start = grant.start;
     cliff = grant.cliff;
     vesting = grant.vesting;
+    revokable = grant.revokable;
+    burnsOnRevoke = grant.burnsOnRevoke;
 
     vested = vestedTokens(grant, uint64(now));
```
Comment (federicobond) [inline contracts/token/VestedToken.sol]: OpenZeppelin security guidelines recommend against using block timestamps. If we are going to keep it, perhaps it would be good to add a comment explaining why.

Comment (izqui) [inline contracts/token/VestedToken.sol]: I think using block numbers in this case would be worse, this contract needs to support limiting token transferability in long term time periods, where the manipulation a miner can do isn't significant. Assuming ETH block times years in advance seems like a source of problems.

File: contracts/token/VestedToken.sol (hunk 2 above)
Comment (izqui) [inline contracts/token/VestedToken.sol]: Attempted a LISP kind of semantics to make the operation more clear. Slipping a parenthesis would be pretty bad.

File: contracts/token/VestedToken.sol (hunk 1 above)
Comment (izqui) [inline contracts/token/VestedToken.sol]: yup, changing this

File: contracts/token/VestedToken.sol (hunk 2 above)
Comment (maraoz) [inline contracts/token/VestedToken.sol]: LGTM

File: contracts/token/VestedToken.sol (hunk 3 above)
Comment (maraoz) [inline contracts/token/VestedToken.sol]: Agreed with @izqui -- thanks @federicobond for pointing that out, though

**Comment by maraoz** [comment]
Just opened a "rebase" PR to your branch: https://github.com/aragon/zeppelin-solidity/pull/2

**Comment by maraoz** [comment]
I did a non-rebase PR because rebasing led to too many conflicts on too many files on each commit.

**Comment by maraoz** [comment]
Thanks @izqui !
